### PR TITLE
:sparkles: Allow for envFrom per container instead of exclusively global

### PR DESCRIPTION
**Pull request by danquack**
#### What this PR does / why we need it:
Each container will be able to specify environment variables from a secret or configmap exclusively, therefore not requiring every container to share environment. When you apply the helm config, the dictionaries of `container.` and `datadog.envFrom` will be merged.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Special notes for your reviewer:
I am happy to update the changelog, but I wasnt sure on the versioning, for when this gets approved as to avoid conflicts with other PRs.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Documentation has been updated with helm-docs (run: `.github/helm-docs.sh`)
- [x] Chart Version bumped
- [x] `CHANGELOG.md` has beed updated
- [x] Variables are documented in the `README.md`

### CODE REVIEW

File: charts/datadog/values.yaml [hunk 1]
```diff
@@ -942,6 +956,9 @@ agents:
       # agents.containers.systemProbe.env -- Additional environment variables for the system-probe container
       env: []
 
+      # agents.containers.systemProbe.envFrom -- Additional environment variables for the system-probe container
+      envFrom: []
```
Comment (L3n41c) [inline charts/datadog/values.yaml:960]: ditto
```suggestion
      # agents.containers.systemProbe.envFrom -- Set environment variables for the system-probe directly from configMaps and/or secrets
      ## envFrom to pass configmaps or secrets as environment
      envFrom: []
      #   - configMapRef:
      #       name: <CONFIGMAP_NAME>
      #   - secretRef:
      #       name: <SECRET_NAME>
```

File: charts/datadog/values.yaml [hunk 2]
```diff
@@ -968,6 +985,9 @@ agents:
       # agents.containers.securityAgent.env -- Additional environment variables for the security-agent container
       env:
 
+      # agents.containers.securityAgent.envFrom -- Additional environment variables for the security-agent container
+      envFrom: []
```
Comment (L3n41c) [inline charts/datadog/values.yaml:989]: ditto
```suggestion
      # agents.containers.securityAgent.envFrom -- Set environment variables for the security-agent directly from configMaps and/or secrets
      ## envFrom to pass configmaps or secrets as environment
      envFrom: []
      #   - configMapRef:
      #       name: <CONFIGMAP_NAME>
      #   - secretRef:
      #       name: <SECRET_NAME>
```

File: charts/datadog/values.yaml [hunk 3]
```diff
@@ -1226,6 +1246,17 @@ clusterChecksRunner:
   #   - name: <ENV_VAR_NAME>
   #     value: <ENV_VAR_VALUE>
 
+  # clusterChecksRunner.envFrom -- Set environment variables for all Agents directly from configMaps and/or secrets
```
Comment (L3n41c) [inline charts/datadog/values.yaml:1249]:
```suggestion
  # clusterChecksRunner.envFrom -- Set environment variables for the Cluster Checks Runner directly from configMaps and/or secrets
```

File: charts/datadog/templates/_container-process-agent.yaml [hunk 4]
```diff
@@ -18,9 +18,9 @@
 {{- end }}
   resources:
 {{ toYaml .Values.agents.containers.processAgent.resources | indent 4 }}
-{{- if .Values.datadog.envFrom }}
+{{- if or (.Values.datadog.envFrom) (.Values.agents.containers.processAgent.envFrom)  }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-process-agent.yaml:21]: ditto
```suggestion
{{- if or .Values.datadog.envFrom .Values.agents.containers.processAgent.envFrom }}
```

File: charts/datadog/templates/_container-process-agent.yaml [hunk 5]
```diff
@@ -18,9 +18,9 @@
...
   envFrom:
-{{ toYaml .Values.datadog.envFrom | indent 4 }}
+{{ toYaml (concat .Values.datadog.envFrom .Values.agents.containers.processAgent.envFrom) | indent 4 }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-process-agent.yaml:23]: ditto
```suggestion
{{ concat .Values.datadog.envFrom .Values.agents.containers.processAgent.envFrom | toYaml | indent 4 }}
```

File: charts/datadog/templates/_container-security-agent.yaml [hunk 6]
```diff
@@ -12,9 +12,9 @@
   ports:
 {{ toYaml .Values.agents.containers.securityAgent.ports | indent 2 }}
 {{- end }}
-{{- if .Values.datadog.envFrom }}
+{{- if or (.Values.datadog.envFrom) (.Values.agents.containers.securityAgent.envFrom)  }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-security-agent.yaml:15]: ditto
```suggestion
{{- if or .Values.datadog.envFrom .Values.agents.containers.securityAgent.envFrom }}
```

File: charts/datadog/templates/_container-security-agent.yaml [hunk 7]
```diff
@@ -12,9 +12,9 @@
...
   envFrom:
-{{ toYaml .Values.datadog.envFrom | indent 4 }}
+{{ toYaml (concat .Values.datadog.envFrom .Values.agents.containers.securityAgent.envFrom) | indent 4 }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-security-agent.yaml:17]: ditto
```suggestion
{{ concat .Values.datadog.envFrom .Values.agents.containers.securityAgent.envFrom | toYaml | indent 4 }}
```

File: charts/datadog/templates/_container-system-probe.yaml [hunk 8]
```diff
@@ -9,9 +9,9 @@
   ports:
 {{ toYaml .Values.agents.containers.systemProbe.ports | indent 2 }}
 {{- end }}
-{{- if .Values.datadog.envFrom }}
+{{- if or (.Values.datadog.envFrom) (.Values.agents.containers.systemProbe.envFrom) }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-system-probe.yaml:12]: ditto
```suggestion
{{- if or .Values.datadog.envFrom .Values.agents.containers.systemProbe.envFrom }}
```

File: charts/datadog/templates/_container-system-probe.yaml [hunk 9]
```diff
@@ -9,9 +9,9 @@
...
   envFrom:
-{{ toYaml .Values.datadog.envFrom | indent 4 }}
+{{ toYaml (concat .Values.datadog.envFrom .Values.agents.containers.systemProbe.envFrom) | indent 4 }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-system-probe.yaml:14]: ditto
```suggestion
{{ concat .Values.datadog.envFrom .Values.agents.containers.systemProbe.envFrom | toYaml | indent 4 }}
```

File: charts/datadog/templates/_container-trace-agent.yaml [hunk 10]
```diff
@@ -22,9 +22,9 @@
 {{- if .Values.agents.containers.traceAgent.ports }}
 {{ toYaml .Values.agents.containers.traceAgent.ports | indent 2 }}
 {{- end }}
-{{- if .Values.datadog.envFrom }}
+{{- if or (.Values.datadog.envFrom) (.Values.agents.containers.traceAgent.envFrom) }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-trace-agent.yaml:25]: ditto
```suggestion
{{- if or .Values.datadog.envFrom .Values.agents.containers.traceAgent.envFrom }}
```

File: charts/datadog/templates/_container-trace-agent.yaml [hunk 11]
```diff
@@ -22,9 +22,9 @@
...
   envFrom:
-{{ toYaml .Values.datadog.envFrom | indent 4 }}
+{{ toYaml (concat .Values.datadog.envFrom .Values.agents.containers.traceAgent.envFrom) | indent 4 }}
```
Comment (L3n41c) [inline charts/datadog/templates/_container-trace-agent.yaml:27]: ditto
```suggestion
{{ concat .Values.datadog.envFrom .Values.agents.containers.traceAgent.envFrom | toYaml | indent 4 }}
```

File: charts/datadog/templates/agent-clusterchecks-deployment.yaml [hunk 12]
```diff
@@ -99,9 +99,9 @@ spec:
         ports:
 {{ toYaml .Values.clusterChecksRunner.ports | indent 10 }}
 {{- end }}
-{{- if .Values.datadog.envFrom }}
+{{- if or (.Values.datadog.envFrom) (.Values.clusterChecksRunner.envFrom)  }}
```
Comment (L3n41c) [inline charts/datadog/templates/agent-clusterchecks-deployment.yaml:102]: ditto
```suggestion
{{- if or .Values.datadog.envFrom .Values.clusterChecksRunner.envFrom }}
```

File: charts/datadog/templates/agent-clusterchecks-deployment.yaml [hunk 13]
```diff
@@ -99,9 +99,9 @@ spec:
...
         envFrom:
-{{ toYaml .Values.datadog.envFrom | indent 10 }}
+{{ toYaml (concat .Values.datadog.envFrom .Values.clusterChecksRunner.envFrom) | indent 10 }}
```
Comment (L3n41c) [inline charts/datadog/templates/agent-clusterchecks-deployment.yaml:104]: ditto
```suggestion
{{ concat .Values.datadog.envFrom .Values.clusterChecksRunner.envFrom | toYaml | indent 10 }}
```

File: charts/datadog/templates/cluster-agent-deployment.yaml [hunk 14]
```diff
@@ -115,9 +115,9 @@ spec:
           name: metricsapi
           protocol: TCP
         {{- end }}
-{{- if .Values.datadog.envFrom }}
+{{- if or (.Values.datadog.envFrom) (.Values.clusterAgent.envFrom) }}
```
Comment (L3n41c) [inline charts/datadog/templates/cluster-agent-deployment.yaml:118]: ditto
```suggestion
{{- if or .Values.datadog.envFrom .Values.clusterAgent.envFrom }}
```

File: charts/datadog/templates/cluster-agent-deployment.yaml [hunk 15]
```diff
@@ -115,9 +115,9 @@ spec:
...
         envFrom:
-{{ toYaml .Values.datadog.envFrom | indent 10 }}
+{{ toYaml (concat .Values.datadog.envFrom .Values.clusterAgent.envFrom) | indent 10 }}
```
Comment (L3n41c) [inline charts/datadog/templates/cluster-agent-deployment.yaml:120]: ditto
```suggestion
{{ concat .Values.datadog.envFrom .Values.clusterAgent.envFrom | toYaml | indent 10 }}
```

**Review by L3n41c — COMMENTED** [review]
Thank you @danquack for this contribution !

I left very minor remarks concerning the coding style.

One very subtle issue that I fixed in my suggestion is that some of the `if` statements were finishing with two spaces instead one.

I.e.:
```
{{- if or […].envFrom  }}
```
instead of
```
{{-if or […].envFrom }}
```

Such two spaces are known to trigger a bug in the go template parser of an old GO version that was used to compile some versions of Helm 2 that are still used in the wild.
So, for Helm 2 compatibility, it’s important to check that mustache directives are not ending with two spaces.

Apart from that, I think the documentation of the new `envFrom` settings should highlight the difference with the already existing `env` ones.

And let’s not forget to bump the chart version and add an entry in the `CHANGELOG.md` when we’ll be close enough to the merge to not fear merge conflicts.

**Comment by danquack** [comment]
@L3n41c thanks for the review. I believe i address all your comments. Looks like there was an `env` culprit too with the double space, so I fixed that as well. Lemme know if there's anything else needed.

**Review by L3n41c — APPROVED** [review]

File: charts/datadog/values.yaml [hunk 16]
```diff
@@ -1226,6 +1270,17 @@ clusterChecksRunner:
...
+  ## envFrom to pass configmaps or secrets as environment
+  ## ref: https://github.com/DataDog/datadog-agent/tree/main/Dockerfiles/agent#environment-variables
+  envFrom: []
+  #   - configMapRef:
+  #       name: <CONFIGMAP_NAME>
+  #   - secretRef:
+  #       name: <SECRET_NAME>
+
+
+
```
Comment (L3n41c) [inline charts/datadog/values.yaml:1283]: [Those extra blank lines are breaking the CI](https://github.com/DataDog/helm-charts/pull/393/checks?check_run_id=3700936167#step:5:48).
```suggestion
```

**Review by L3n41c — COMMENTED** [review]
The CI is broken by some blank lines.
I think we can merge as soon as the CI is fixed.

**Comment by danquack** [comment]
Looks like theres something external to this change. Not sure how these are handled:
```
 ✖︎ synthetics-private-location => (version: "0.8.0", path: "charts/synthetics-private-location") > Chart version not ok. Needs a version bump!
 ```

**Pull request merged**